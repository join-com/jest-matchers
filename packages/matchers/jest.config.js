module.exports = {
  testEnvironment: 'node',
  testPathIgnorePatterns: ['/node_modules/'],
  transform: { '^.+\\.tsx?$': ['ts-jest', { diagnostics: false }] },
  maxWorkers: 1,
  workerIdleMemoryLimit: '1.5GB',
}
