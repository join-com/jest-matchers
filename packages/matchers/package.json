{
  "name": "@join-private/jest-matchers",
  "version": "0.0.0-development",
  "description": "Set of custom jest matchers",
  "author": "JOIN Solutions AG",
  "license": "UNLICENSED",
  "main": "lib/index.js",
  "types": "types/index.d.ts",
  "files": [
    "lib",
    "types/index.d.ts"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/join-com/jest-matchers"
  },
  "scripts": {
    "build": "tsc -p tsconfig.prod.json",
    "test": "jest",
    "test:ci": "jest --no-color",
    "test:watch": "jest --watch",
    "lint": "yarn lint:tsc && yarn lint:eslint",
    "lint:eslint": "eslint . --ext .ts --max-warnings 0",
    "lint:tsc": "tsc --noEmit",
    "prepublishOnly": "yarn lint && yarn build"
  },
  "keywords": [
    "jest",
    "matchers"
  ],
  "engines": {
    "node": ">=14.0.0"
  },
  "devDependencies": {
    "@join-private/eslint-config-backend": "^1.5.0",
    "@types/expect": "^24.3.0",
    "@types/jest": "^29.5.12",
    "@types/node": "^22.5.4",
    "eslint": "^8.18.0",
    "jest": "^29.7.0",
    "prettier": "^3.3.3",
    "ts-jest": "^29.2.5",
    "typescript": "^5.5.4"
  },
  "dependencies": {
    "expect": "^29.7.0",
    "fast-deep-equal": "^3.1.3",
    "jest-matcher-utils": "29.7.0"
  },
  "publishConfig": {
    "access": "restricted",
    "registry": "https://europe-west1-npm.pkg.dev/join-dev-330413/join-private/"
  }
}
