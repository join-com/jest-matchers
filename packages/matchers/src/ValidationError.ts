export interface IFieldError {
  fieldName: string
  message: string
  constraint: string
}

export class ValidationError extends Error {
  public readonly type = 'ApplicationError'
  public readonly code = 'validation'

  constructor(
    message: string,
    public readonly fields: IFieldError[],
  ) {
    super(message)
  }
}
