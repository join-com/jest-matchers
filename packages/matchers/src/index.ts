/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import * as matchers from './matchers'

const jestExpect = global['expect']

if (jestExpect !== undefined) {
  jestExpect.extend(matchers)
} else {
  console.error(`
    Unable to find Jest's global expect.
    Please check you have added jest-matchers to setupFilesAfterEnv section of jest config
  `)
}
