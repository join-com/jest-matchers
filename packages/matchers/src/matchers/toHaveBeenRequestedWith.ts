import { diff, matcherHint, printExpected, printReceived, RECEIVED_COLOR } from 'jest-matcher-utils'

type ICallWithRequest = [{ request: Record<string, unknown> }]

interface IMock {
  mock: {
    calls: ICallWithRequest[]
  }
}

export const toHaveBeenRequestedWith = ({ mock }: IMock, expected: unknown): jest.CustomMatcherResult => {
  const calls = mock.calls
  const pass = calls.length > 0 && anyOfRequestsSucceeded(calls, expected)
  return {
    pass,
    message: () => {
      const hintString = matcherHint('.toHaveBeenRequestedWith')
      const expectedString = printExpected(expected)
      const mismatchString = formatMismatchedCalls(mock.calls, expected)
      return `${hintString}\n\n` + `Expected request value to equal:\n${expectedString}\n` + mismatchString
    },
  }
}

const formatMismatchedCalls = (calls: ICallWithRequest[], expected: unknown): string => {
  if (!calls.length) {
    return `But it was ${RECEIVED_COLOR('not called')}.`
  }

  const messages: string[] = ['Received:']
  calls.map((call, index) => {
    const received = call[0].request
    const diffString = diff(expected, received) ?? ''

    const receivedString = printReceived(received)
    const callNumber: string = calls.length > 1 ? `\ncall no.${String(index + 1)}\n` : ''
    const callMessage = `${callNumber}` + `${receivedString}\n` + 'Difference:\n' + `${diffString}`

    messages.push(callMessage)
  })

  if (calls.length > 1) {
    messages.push(`\nNumber of calls: ${calls.length}`)
  }
  return messages.join('\n')
}

const anyOfRequestsSucceeded = (calls: ICallWithRequest[], expected: unknown): boolean => {
  for (const call of calls) {
    if (equal(call[0].request, expected)) {
      return true
    }
  }
  return false
}

const equal = (request: Record<string, unknown>, expected: unknown): boolean => {
  try {
    expect(request).toEqual(expected)
    return true
  } catch (e) {
    return false
  }
}
