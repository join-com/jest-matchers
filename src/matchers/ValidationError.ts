export class ValidationError extends Error {
  public readonly type = 'ApplicationError';
  public readonly code = 'validation';

  constructor(
    message: string,
    public readonly fields: {
      fieldName: string;
      message: string;
      constraint: string;
    }[],
  ) {
    super(message);
  }
}
