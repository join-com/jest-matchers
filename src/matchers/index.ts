export * from './toHaveBeenRequestedWith';
export * from './toThrowValidationError';
