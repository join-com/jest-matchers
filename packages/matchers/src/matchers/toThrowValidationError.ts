import equal from 'fast-deep-equal'
import { printDiffOrStringify } from 'jest-matcher-utils'
import { ValidationError } from '../ValidationError'

function getThrown(received: Error | (() => void)): Error | undefined {
  if (received instanceof Error) {
    return received
  }

  try {
    received()
  } catch (error) {
    return error as Error
  }

  return
}

export function toThrowValidationError<T extends ValidationError>(
  received: Error | (() => void),
  expected: T,
): jest.CustomMatcherResult {
  const thrown = getThrown(received)
  if (!thrown) {
    return {
      pass: false,
      message: () => 'Expected to receive an error or a function that throws an error',
    }
  }

  const isExpectedClass = thrown instanceof expected.constructor
  if (!isExpectedClass) {
    return {
      pass: false,
      message: () =>
        `Expected error to be an instance of ${expected.constructor.name} but got ${thrown.constructor.name}`,
    }
  }

  const receivedFields = (thrown as ValidationError).fields
  const areFieldsMatching = equal(receivedFields, expected.fields)
  if (!areFieldsMatching) {
    return {
      pass: false,
      message: () => printDiffOrStringify(receivedFields, expected.fields, 'Expected fields', 'Received fields', true),
    }
  }

  return { pass: true, message: () => '' }
}
