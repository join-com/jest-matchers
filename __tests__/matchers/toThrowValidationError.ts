import { ValidationError } from '../../src/matchers/ValidationError';
import { toThrowValidationError } from '../../src/matchers/toThrowValidationError';

expect.extend({ toThrowValidationError });

describe('toThrowValidationError', () => {
  const expectedValidationError = () =>
    new ValidationError('Expected validation error', [
      {
        fieldName: 'someField',
        constraint: 'mustBeCorrect',
        message: 'Some field must be correct but its not',
      },
    ]);

  it('passess for function throwing expected ValidationError', () => {
    const fn = () => {
      throw expectedValidationError();
    };

    expect(fn).toThrowValidationError(expectedValidationError());
  });

  it('passess for directly passing the expected ValidationError', () => {
    expect(expectedValidationError()).toThrowValidationError(
      expectedValidationError(),
    );
  });

  it('fails when error is not thrown', () => {
    const fn = () => {};

    const validate = () => {
      expect(fn).toThrowValidationError(expectedValidationError());
    };
    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('fails for error of other type', () => {
    const fn = () => {
      throw new Error('other type');
    };

    const validate = () => {
      expect(fn).toThrowValidationError(expectedValidationError());
    };
    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('fails for ValidationError with different fieldName', () => {
    const fn = () => {
      const error = expectedValidationError();
      error.fields[0].fieldName = 'different';
      throw error;
    };

    const validate = () => {
      expect(fn).toThrowValidationError(expectedValidationError());
    };
    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('fails for ValidationError with different constraint', () => {
    const fn = () => {
      const error = expectedValidationError();
      error.fields[0].constraint = 'different';
      throw error;
    };

    const validate = () => {
      expect(fn).toThrowValidationError(expectedValidationError());
    };
    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('fails for ValidationError with different message', () => {
    const fn = () => {
      const error = expectedValidationError();
      error.fields[0].message = 'different';
      throw error;
    };

    const validate = () => {
      expect(fn).toThrowValidationError(expectedValidationError());
    };
    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('fails for ValidationError with additional fields', () => {
    const fn = () => {
      const error = expectedValidationError();
      error.fields[1] = {
        fieldName: 'someOtherField',
        constraint: 'mustBeCorrectToo',
        message: 'Some other field must be correct but its not',
      };
      throw error;
    };

    const validate = () => {
      expect(fn).toThrowValidationError(expectedValidationError());
    };
    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('fails for ValidationError with no fields', () => {
    const fn = () => {
      const error = expectedValidationError();
      error.fields.pop();
      throw error;
    };

    const validate = () => {
      expect(fn).toThrowValidationError(expectedValidationError());
    };
    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('fails for ValidationError with different fields', () => {
    const fn = () => {
      const error = expectedValidationError();
      error.fields[0] = {
        fieldName: 'someOtherField',
        constraint: 'mustBeCorrectToo',
        message: 'Some other field must be correct but its not',
      };
      throw error;
    };

    const validate = () => {
      expect(fn).toThrowValidationError(expectedValidationError());
    };
    expect(validate).toThrowErrorMatchingSnapshot();
  });
});
