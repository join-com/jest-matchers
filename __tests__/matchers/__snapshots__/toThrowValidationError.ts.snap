// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toThrowValidationError fails for ValidationError with additional fields 1`] = `
"[32m- Expected fields  - 5[39m
[31m+ Received fields  + 0[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"constraint\\": \\"mustBeCorrect\\",[22m
[2m      \\"fieldName\\": \\"someField\\",[22m
[2m      \\"message\\": \\"Some field must be correct but its not\\",[22m
[2m    },[22m
[32m-   Object {[39m
[32m-     \\"constraint\\": \\"mustBeCorrectToo\\",[39m
[32m-     \\"fieldName\\": \\"someOtherField\\",[39m
[32m-     \\"message\\": \\"Some other field must be correct but its not\\",[39m
[32m-   },[39m
[2m  ][22m"
`;

exports[`toThrowValidationError fails for ValidationError with different constraint 1`] = `
"[32m- Expected fields  - 1[39m
[31m+ Received fields  + 1[39m

[2m  Array [[22m
[2m    Object {[22m
[32m-     \\"constraint\\": \\"different\\",[39m
[31m+     \\"constraint\\": \\"mustBeCorrect\\",[39m
[2m      \\"fieldName\\": \\"someField\\",[22m
[2m      \\"message\\": \\"Some field must be correct but its not\\",[22m
[2m    },[22m
[2m  ][22m"
`;

exports[`toThrowValidationError fails for ValidationError with different fieldName 1`] = `
"[32m- Expected fields  - 1[39m
[31m+ Received fields  + 1[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"constraint\\": \\"mustBeCorrect\\",[22m
[32m-     \\"fieldName\\": \\"different\\",[39m
[31m+     \\"fieldName\\": \\"someField\\",[39m
[2m      \\"message\\": \\"Some field must be correct but its not\\",[22m
[2m    },[22m
[2m  ][22m"
`;

exports[`toThrowValidationError fails for ValidationError with different fields 1`] = `
"[32m- Expected fields  - 3[39m
[31m+ Received fields  + 3[39m

[2m  Array [[22m
[2m    Object {[22m
[32m-     \\"constraint\\": \\"mustBeCorrectToo\\",[39m
[32m-     \\"fieldName\\": \\"someOtherField\\",[39m
[32m-     \\"message\\": \\"Some other field must be correct but its not\\",[39m
[31m+     \\"constraint\\": \\"mustBeCorrect\\",[39m
[31m+     \\"fieldName\\": \\"someField\\",[39m
[31m+     \\"message\\": \\"Some field must be correct but its not\\",[39m
[2m    },[22m
[2m  ][22m"
`;

exports[`toThrowValidationError fails for ValidationError with different message 1`] = `
"[32m- Expected fields  - 1[39m
[31m+ Received fields  + 1[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"constraint\\": \\"mustBeCorrect\\",[22m
[2m      \\"fieldName\\": \\"someField\\",[22m
[32m-     \\"message\\": \\"different\\",[39m
[31m+     \\"message\\": \\"Some field must be correct but its not\\",[39m
[2m    },[22m
[2m  ][22m"
`;

exports[`toThrowValidationError fails for ValidationError with no fields 1`] = `
"[32m- Expected fields  - 1[39m
[31m+ Received fields  + 7[39m

[32m- Array [][39m
[31m+ Array [[39m
[31m+   Object {[39m
[31m+     \\"constraint\\": \\"mustBeCorrect\\",[39m
[31m+     \\"fieldName\\": \\"someField\\",[39m
[31m+     \\"message\\": \\"Some field must be correct but its not\\",[39m
[31m+   },[39m
[31m+ ][39m"
`;

exports[`toThrowValidationError fails for error of other type 1`] = `"Expected error to be an instance of ValidationError but got Error"`;

exports[`toThrowValidationError fails when error is not thrown 1`] = `"Expected to receive an error or a function that throws an error"`;
